### PR TITLE
suggest iter_mut() where trying to modify elements from .iter()

### LINKED ISSUE #62387: diagnostics: suggest iter_mut() where trying to modify .iter()'ed vector elements inplace
code:
````rust
#[derive(Debug)]
struct A {
    a: i32,
}

impl A {
    fn double(&mut self) {
        self.a += self.a
    }
}

fn main() {
    let mut v = [A { a: 4 }];
    v.iter().for_each(|a| a.double());
    println!("{:?}", v);
}
````

The code gives the following not-very-helpful warning:
````
error[E0596]: cannot borrow `*a` as mutable, as it is behind a `&` reference
  --> src/main.rs:14:27
   |
14 |     v.iter().for_each(|a| a.double());
   |                        -  ^ `a` is a `&` reference, so the data it refers to cannot be borrowed as mutable
   |                        |
   |                        help: consider changing this to be a mutable reference: `&mut A`

warning: variable does not need to be mutable
  --> src/main.rs:13:9
   |
13 |     let mut v = [A { a: 4 }];
   |         ----^
   |         |
   |         help: remove this `mut`
   |
   = note: #[warn(unused_mut)] on by default

error: aborting due to previous error
````

trying to use
````rust
v.iter().for_each(|&mut a| a.double());
````
and a none-mutable vector just causes more errors
````
error[E0308]: mismatched types
  --> src/main.rs:14:24
   |
14 |     v.iter().for_each(|&mut a| a.double());
   |                        ^^^^^^ types differ in mutability
   |
   = note: expected type `&A`
              found type `&mut _`
   = help: did you mean `mut a: &&A`?
````

The actual fix is to use `iter_mut()` instead of just `iter()`:

````rust
#[derive(Debug)]
struct A {
    a: i32,
}

impl A {
    fn double(&mut self) {
        self.a += self.a
    }
}

fn main() {
    let mut v = [A { a: 4 }];
    v.iter_mut().for_each(|a| a.double());
    println!("{:?}", v);
}
````

It would be very helpful the compiler could suggest `iter_mut()`!

### LINKED ISSUE #115259: Wrong syntax suggested by "consider changing this to be a mutable reference" with mut parameter in closure when iterating a vector
### Code

```Rust
fn main() {}

pub trait Layer {
	fn process(&mut self) -> u32;
}

pub struct State {
	layers: Vec<Box<dyn Layer>>,
}

impl State {
	pub fn process(&mut self) -> u32 {
		self.layers.iter().fold(0, |result, mut layer| result + layer.process())
	}
}
```

### Current output

```Shell
error[E0596]: cannot borrow `**layer` as mutable, as it is behind a `&` reference
  --> src/main.rs:13:59
   |
13 |         self.layers.iter().fold(0, |result, mut layer| result + layer.process())
   |                                                                 ^^^^^^^^^^^^^^^ `layer` is a `&` reference, so the data it refers to cannot be borrowed as mutable
   |
help: consider changing this to be a mutable reference
   |
13 |         self.layers.iter().fold(0, |result, mut lmut ayer| result + layer.process())
   |                                                  +++
```

### Desired output

_No response_

### Rationale and extra context

Initially, my `State.process` function looked like this:

```rust
self.layers.iter().fold(0, |result, layer| result + layer.process())
```

This led to a diagnostic that already has an open issue #62387 (it should suggest `iter` -> `iter_mut`).

I didn't notice I was not using `iter_mut` and tried changing `layer` -> `mut layer`, which led to the compiler somehow adding a second `mut` in the middle of the closure parameter.

obrazek

I don't know what the desired output should be in this case; I don't think `mut` parameter is correct here either way, but it might be worth investigating why the compiler is making a syntactically wrong suggestion.

### Other cases

_No response_

### Anything else?

Tested on 1.71.0 and 1.72.0 (stable-x86_64-unknown-linux-gnu).

### PR DESCRIPTION
**Pull request by chenyukang**
Fixes https://github.com/rust-lang/rust/issues/115259
Fixes https://github.com/rust-lang/rust/issues/62387

### CODE REVIEW

File: compiler/rustc_borrowck/src/diagnostics/mutability_errors.rs [hunk 1]
```diff
@@ -953,6 +955,41 @@ impl<'a, 'tcx> MirBorrowckCtxt<'a, 'tcx> {
         }
     }
 
+    fn suggest_using_iter_mut(&self, err: &mut DiagnosticBuilder<'_, ErrorGuaranteed>) {
+        let source = self.body.source;
+        let hir = self.infcx.tcx.hir();
+        if let InstanceDef::Item(def_id) = source.instance
+            && let Some(Node::Expr( hir::Expr {hir_id, kind, ..})) = hir.get_if_local(def_id)
```
Comment (davidtwco) [inline compiler/rustc_borrowck/src/diagnostics/mutability_errors.rs:962]:
```suggestion
            && let Some(Node::Expr(hir::Expr { hir_id, kind, ..})) = hir.get_if_local(def_id)
```
nit: extraneous space and lack of a space

File: compiler/rustc_borrowck/src/diagnostics/mutability_errors.rs [hunk 2]
```diff
@@ -953,6 +955,41 @@ impl<'a, 'tcx> MirBorrowckCtxt<'a, 'tcx> {
...
+            && let ExprKind::Closure(closure) = kind
+            && closure.movability == None
+            && let Some(Node::Expr(expr)) = hir.find_parent(*hir_id) {
+                let mut cur_expr = expr;
+                loop {
+                    if let ExprKind::MethodCall(path_segment, recv, _, _) = cur_expr.kind {
```
Comment (davidtwco) [inline compiler/rustc_borrowck/src/diagnostics/mutability_errors.rs:968]: can a `while let` be used here instead?

File: compiler/rustc_borrowck/src/diagnostics/mutability_errors.rs [hunk 3]
```diff
@@ -953,6 +955,41 @@ impl<'a, 'tcx> MirBorrowckCtxt<'a, 'tcx> {
...
+                        if path_segment.ident.name == sym::iter {
+                            let def_id = hir.enclosing_body_owner(recv.hir_id);
+                            let tables = self.infcx.tcx.typeck(def_id);
+                            let self_ty = tables.node_type_opt(recv.hir_id);
+                            if let Some(_ty) = self_ty {
+                                // FIXME: check `_ty` has `iter_mut` method ?
```
Comment (davidtwco) [inline compiler/rustc_borrowck/src/diagnostics/mutability_errors.rs:974]: This shouldn't be too tricky to do - [here's an example of Esteban doing something similar in a recent PR of his](https://github.com/rust-lang/rust/pull/115147/files#diff-23edf23f7466b97eb3154b82f7b324983287c327937fd66e854b3df6759926ceR817) (from my recollection of reviewing it, didn't spend too much time confirming that it is what you need, but hopefully a starting point).

File: tests/ui/borrowck/issue-62387-suggest-iter-mut.rs [hunk 4]
```diff
@@ -0,0 +1,82 @@
...
+
+fn bar() {
+    let mut v = [A { a: 4 }];
+    v.iter().rev().rev().for_each(|a| a.double());
+    //~^ ERROR cannot borrow `*a` as mutable, as it is behind a `&` reference
+    println!("{:?}", v);
+}
+
+// --------------------------------------------------------------------------
+// test case from #115259
```
Comment (davidtwco) [inline tests/ui/borrowck/issue-62387-suggest-iter-mut.rs:32]: You can just make these different files if you want.

**Review by davidtwco — CHANGES_REQUESTED** [review]
I like this, left a few comments.

File: compiler/rustc_borrowck/src/diagnostics/mutability_errors.rs (hunk 2 above)
Comment (chenyukang) [reply compiler/rustc_borrowck/src/diagnostics/mutability_errors.rs]: ah, great point, fixed the code.

File: compiler/rustc_borrowck/src/diagnostics/mutability_errors.rs (hunk 3 above)
Comment (chenyukang) [reply compiler/rustc_borrowck/src/diagnostics/mutability_errors.rs]: nice!
thanks for the tip, updated the code.

File: compiler/rustc_borrowck/src/diagnostics/mutability_errors.rs [hunk 5]
```diff
@@ -953,6 +956,44 @@ impl<'a, 'tcx> MirBorrowckCtxt<'a, 'tcx> {
...
+                            .type_dependent_def_id(cur_expr.hir_id)
+                            .and_then(|def_id| self.infcx.tcx.impl_of_method(def_id))
+                            .map(|def_id| self.infcx.tcx.associated_items(def_id))
+                            .map(|assoc_items| {
+                                assoc_items.filter_by_name_unhygienic(sym::iter_mut).peekable()
+                            });
+
+                        if let Some(mut res) = res && res.peek().is_some() {
+                            err.span_suggestion_verbose(
+                            path_segment.ident.span,
```
Comment (davidtwco) [inline compiler/rustc_borrowck/src/diagnostics/mutability_errors.rs:983]: nit: fmt

File: compiler/rustc_borrowck/src/diagnostics/mutability_errors.rs [hunk 6]
```diff
@@ -810,6 +812,7 @@ impl<'a, 'tcx> MirBorrowckCtxt<'a, 'tcx> {
             };
             v.visit_block(block);
             if let Some(expr) = v.expr && let Call(_, [expr]) = expr.kind {
+                debug!("anan expr: {:#?}", expr);
```
Comment (davidtwco) [inline compiler/rustc_borrowck/src/diagnostics/mutability_errors.rs:815]: anan?

**Review by davidtwco — APPROVED** [review]
LGTM, two small nits

File: compiler/rustc_borrowck/src/diagnostics/mutability_errors.rs (hunk 6 above)
Comment (chenyukang) [reply compiler/rustc_borrowck/src/diagnostics/mutability_errors.rs]: ah, forgot to remove my debug code.
clean it.

**Review by davidtwco — APPROVED** [review]

**Pull request merged**